internal/service/streamprocessor/state_transition_test.go

```diff
@@ -0,0 +1,156 @@
+package streamprocessor_test
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"testing"
+
+	"go.mongodb.org/atlas-sdk/v20240805001/admin"
+	"go.mongodb.org/atlas-sdk/v20240805001/mockadmin"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/service/streamprocessor"
+)
+
+var (
+	InitiatingState     = "INIT"
+	CreatingState       = "CREATING"
+	CreatedState        = "CREATED"
+	StartedState        = "STARTED"
+	StoppedState        = "STOPPED"
+	DroppedState        = "DROPPED"
+	FailedState         = "FAILED"
+	sc500               = conversion.IntPtr(500)
+	sc200               = conversion.IntPtr(200)
+	sc404               = conversion.IntPtr(404)
+	streamProcessorName = "processorName"
+	requestParams       = &admin.GetStreamProcessorApiParams{
+		GroupId:       "groupId",
+		TenantName:    "tenantName",
+		ProcessorName: streamProcessorName,
+	}
+)
+
+type testCase struct {
+	expectedState *string
+	name          string
+	mockResponses []response
+	desiredStates []string
+	pendingStates []string
+	expectedError bool
+}
+
+func TestStreamProcessorStateTransition(t *testing.T) {
+	testCases := []testCase{
+		{
+			name: "Successful transition to CREATED",
+			mockResponses: []response{
+				{state: &InitiatingState, statusCode: sc200},
+				{state: &CreatingState, statusCode: sc200},
+				{state: &CreatedState, statusCode: sc200},
+			},
+			expectedState: &CreatedState,
+			expectedError: false,
+			desiredStates: []string{CreatedState, FailedState},
+			pendingStates: []string{InitiatingState, CreatingState},
+		},
+		{
+			name: "Successful transition to STARTED",
+			mockResponses: []response{
+				{state: &CreatedState, statusCode: sc200},
+				{state: &StartedState, statusCode: sc200},
+			},
+			expectedState: &StartedState,
+			expectedError: false,
+			desiredStates: []string{StartedState},
+			pendingStates: []string{CreatedState, StoppedState},
+		},
+		{
+			name: "Successful transition to STOPPED",
+			mockResponses: []response{
+				{state: &StartedState, statusCode: sc200},
+				{state: &StoppedState, statusCode: sc200},
+			},
+			expectedState: &StoppedState,
+			expectedError: false,
+			desiredStates: []string{StoppedState},
+			pendingStates: []string{StartedState},
+		},
+		{
+			name: "Error when transitioning to FAILED state",
+			mockResponses: []response{
+				{state: &InitiatingState, statusCode: sc200},
+				{state: &FailedState, statusCode: sc200},
+			},
+			expectedState: nil,
+			expectedError: true,
+			desiredStates: []string{CreatedState, FailedState},
```

Comment:
[nit] I don't think `FailedState` should be part of targets. From the docs of `WaitForStateContext`:

> If the Refresh function returns an error, exit immediately with that error.

Reply:
Removed in [1e09f4e](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2501/commits/1e09f4e3d4f54ad618499d2a7a71cb8c1237533c)